[자료구조] 3주차 Woostagram 기능 구현

os94 opened the pull request:
자료구조 3주차 Wootstagram 기능 추가 목록입니다!
- 사용자간 DM(채팅) 기능 추가
- 댓글 기능 추가
- 팔로우-팔로워 기능 추가
- 해쉬태그 검색 및 결과화면 추가
- 좋아요 기능 추가
감사합니다!

os94 committed:
feat : WebSocket을 이용한 sample 채팅기능 추가

jinwookh committed:
refactor: db경로에 utf8 설정 추가 (#26)

good

ep1stas1s committed:
feat: 해시태그 검색 기능 (#29)

* refactor: 불필요한 코드 제거

* refactor: 불필요한 class 제거

* feat: article 수정 시 hash tag 수정

* feat: hash tag 검색창 수정

* feat: hash tag 검색 시 게시물 수 출력

* fix: 중복 문자열 인식

* chore: 초기 데이터 추가

* refactor: template 수정

* feat: hash tag로 article 검색 (page 적용)

* chore: 초기 data 수정

* refactor: 중복 변수 처리

* feat: hash tag 검색 기능 추가 (무한 스크롤 적용)

* fix: 중복 id 사용 제거

* chore: 초기 data 제거

* fix: 한글 hash tag

* feat: 게시물 수 추가

- 디자인 변경

* refactor: template 분리

* refactor: js파일 script.html (layout)으로 이동

* fix: 검색 조회 쿼리 수정

imkimheejoo committed:
MyFeed 기능 구현 (#31)

* feat: myFeed 기능 구현

* fix: 회원수정시 session에 수정된 loggedIn 유저로 변경

* feat: 다른사람의 myFeed front 구현

* fix: myfeed 사용자 정보 style 수정

* refactor: myFeed 기능 전체 리팩토링

jinwookh committed:
fix: 경로 버그 수정

jinwookh committed:
refactor: 파일 저장 기본 경로를 user.dir -> user.home으로 변경

user.dir는 어플리케이션이 실행되는 파일이 있는 경로를 의미

jinwookh committed:
refactor: 파일 저장 경로를 (날짜 +article/files)를 (article/file/ + 날짜)로 순서 변경

유저 이미지 파일 경로도 동일하게 수정

jinwookh committed:
fix: profile image가 없을 때 null alert 뜨는 버그 수정

profile image가 없으면 서버에서 프런트에게 empty byte를 보내도록 함.

jinwookh committed:
add: circleci config.yml file

jinwookh committed:
add: circleci config.yml file

jinwookh committed:
add: test task at gradle

imkimheejoo committed:
feat: follow 기능 구현 (#36)

* feat: follow 기능 구현

* feat: unfollow 기능 구현

* feat: follwers 목록 modal 구현

* fix: follwers 목록 style 수정

* test: testController 생성

jinwookh committed:
delete: circle ci configurations

os94 committed:
Merge branch 'develop' of https://github.com/wooteco-datastructure/miniprojects-2019 into feat/dm-sean

ep1stas1s committed:
feat: 해시태그 검색 시 캐싱 적용 (#37)

* fix: 테스트 시 data.sql 적용 해제

* feat: hash tag 검색 시 caching 적용

* fix: hash tag regex 수정

os94 committed:
Merge branch 'develop' of https://github.com/wooteco-datastructure/miniprojects-2019 into feat/dm-sean

os94 committed:
refactor : 글 작성 버튼 이동

os94 committed:
feat : WebSocket sample 코드 삭제

os94 committed:
feat : DM기능 추가, 관련 도메인 객체 추가중

os94 committed:
Merge remote-tracking branch 'origin/feat/store_file_refactor' into develop

# Conflicts:
#	src/test/resources/application.properties

os94 committed:
Merge remote-tracking branch 'origin/fix/user_profile_alert_bug' into develop

# Conflicts:
#	src/test/resources/application.properties

imkimheejoo committed:
feat: 팔로우 기능 테스트 및 팔로잉들만 글 조회 기능 (#41)

* feat: follow 기능 구현

* feat: unfollow 기능 구현

* feat: follwers 목록 modal 구현

* fix: follwers 목록 style 수정

* test: testController 생성

* chore: 팀회의 의견 반영

* feat: following list 기능 구현

* test: FeedApiController Test 구현

* test: FeedController Test 구현

* feat: 팔로잉된 사람의 글만 조회 기능 구현

* feat: 팔로잉된 사람의 글만 조회 기능 구현

* fix: 마이페이지 글 목록 순서 날짜 역순으로 수정

ep1stas1s committed:
feat : DM 기능 추가

os94 committed:
Merge remote-tracking branch 'origin/feat/dm' into develop

# Conflicts:
#	src/main/java/com/woowacourse/dsgram/domain/Follow.java
#	src/main/java/com/woowacourse/dsgram/service/FollowService.java
#	src/main/java/com/woowacourse/dsgram/service/facade/Facade.java
#	src/main/java/com/woowacourse/dsgram/web/controller/HomeController.java

os94 committed:
chore : code formatting

jinwookh committed:
Feat/apply profile image (#43)

* delete: circle ci configurations

* feat: infinite scrolling feature in home page

* add: data.sql file

* refactor: index.html에서 필요없는 script 문 삭제

* feat: home에서 유저 프로필 사진 나오도록 기능 추가

* feat: 프로필 이미지가 없으면 기본 이미지로 설정되도록 함.

home만 적용

* feat: article.html페이지에 프로필 이미지가 나오도록 함.

* doc: add todos

* refactor: 무한 스크롤 javascript 함수 중복 제거

search-script에서 무한 스크롤 관련 javascript method가 두 번 적용되는 이슈가 있어서
-> serarch-result.html에만 적용되는 layout 만듬

* refactor: 해시 테그 검색 화면에서 프로필 사진 적용

* refactor: user profile image 가져오는 함수에서 filename 매개변수 제거

* refactor: profile image class에 thumb-img-user-id 설정

* fix: Article 기능 수정

* refactor: user current time() in create_time field at data.sql

also add findAll at ArticleService method

* refactor: 스크립트 import하는 순서 바꿈

* add time stamp field at data.sql

* add: 새로운 사용자, 팔로우 정보 in data.sql

* feat: 해시태그 검색 화면에서 유저 프로필 사진 적용

무한 스크롤 기능 작동함

* feat: 홈 화면에서 유저 프로필 사진 적용

무한 스크롤 기능 작동하지 않음

* add: datas with new user in data.sql

* delete: myFeed.html

* feat: myfeed 페이지에서 article들에 대해 유저 프로필 이미지 적용

* refactor: userProfile image 가져오는 함수 리팩토링

* feat: my feed 화면 상단에 프로필 이미지 적용

무한 스크롤 적용되지 않음

* refactor : BaseEntity 하나로 통합

os94 committed:
fix : resolve merge conflict

os94 committed:
feat : article comment (#44)

* fix: Article 기능 수정

* refactor: comment js

* fix: 댓글 생성, 조회 기능 수정

* fix: 댓글 삭제 기능 수정중

os94 committed:
fix : ArticleAssembler의 merge conflict 해결

os94 committed:
fix : article.html에 comment.js 추가

os94 committed:
fix : my-feed.html에 빠진 follow.js 추가

imkimheejoo committed:
fix: data.sql 수정

os94 committed:
chore : HTML 수정

os94 committed:
Merge branch 'develop' of https://github.com/wooteco-datastructure/miniprojects-2019 into develop

soojinroh committed:
fix: 댓글 기능 수정 (#47)

os94 committed:
fix : 전반적인 코드 정리 (#45)

* refactor : facade 패턴을 사용하지 않는 방식으로 통일

* refactor : ResponseEntity 리턴 방식 통일

* refactor : 전체 리팩토링 및 코드 정리

- 주석 삭제
- 사용하지않는 exception 삭제
- 빠진 @Repository 애노테이션 추가
- 도메인 모델, Dto 등 롬복 사용방식 통일

* refactor : 패키지 구조 수정

* chore : HTML, CSS 수정

- 각 카드사이 배경을 투명하게 수정

jinwookh committed:
Fix/infinite scrolling (#46)

* fix: hash tag 추출 regex 수정

* fix: css 수정

* fix: 검색 게시글 수 기능 원복

* fix: 글쓰기 버튼 삭제

* feat: my feed 무한 스크롤 적용

* feat: 상단 프로필 이미지 적용

* fix: url path 변경 (user -> users)

* fix: profile image 적용

* feat: 최신순으로 정렬

ep1stas1s committed:
Fix/home (#48)

* fix: hash tag 추출 regex 수정

* fix: css 수정

* fix: 검색 게시글 수 기능 원복

* fix: 글쓰기 버튼 삭제

* feat: my feed 무한 스크롤 적용

* feat: 상단 프로필 이미지 적용

* fix: url path 변경 (user -> users)

* fix: profile image 적용

* feat: 최신순으로 정렬

* fix: sort 방식 변경

pobiconan commented on src/main/java/com/woowacourse/dsgram/domain/BaseEntity.java:
```diff
@@ -0,0 +1,24 @@
+@Getter
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public abstract class BaseEntity {
```
Base 객체 추가 👍

pobiconan commented on src/main/java/com/woowacourse/dsgram/domain/BaseEntity.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Column
+    @LastModifiedDate
+    LocalDateTime modifiedDate;
```
이 녀석은 private이 아닌 이유는?

soojinroh replied:
빠뜨린 것 같습니다ㅜㅜ 추가했습니다!

pobiconan commented on src/main/java/com/woowacourse/dsgram/domain/BaseEntity.java:
```diff
@@ -0,0 +1,24 @@
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public abstract class BaseEntity {
+    @Column
```
`@Column` 설정 만으로 끝인가?
예를 들어 생성일은 변경될 수 없으니 `updatable=false`와 같은 설정 추가하면 좋지 않을까?

pobiconan commented on src/main/java/com/woowacourse/dsgram/domain/ChatMessage.java:
```diff
@@ -0,0 +1,44 @@
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+@Getter
+@EqualsAndHashCode(of = {"id"})
+public class ChatMessage extends BaseEntity {
```
수정 기능은 없는 것 같은데 BaseEntity를 통해 `LocalDateTime modifiedDate`도 관리되도록 하는 것이 맞나?

ep1stas1s replied:
반영했습니다^^;

pobiconan commented on src/main/java/com/woowacourse/dsgram/domain/Comment.java:
```diff
@@ -0,0 +1,65 @@
+@NoArgsConstructor
+@Getter
+@EqualsAndHashCode(of = {"id"})
+public class Comment {
```
Comment는 생성일과 수정일 없어도 되나?

soojinroh replied:
BaseEntity를 extends함으로써 생성일과 수정일도 추가하였습니다!

pobiconan commented on src/main/java/com/woowacourse/dsgram/domain/Follow.java:
```diff
@@ -0,0 +1,38 @@
+public class Follow extends BaseEntity {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    long id;
```
인스턴스 변수의 접근 제어자가 private이 아닌 이유는?

imkimheejoo replied:
실수로 접근제어자를 빼먹었습니다. 수정했습니다!

pobiconan commented on src/main/java/com/woowacourse/dsgram/domain/repository/CommentRepository.java:
```diff
@@ -0,0 +1,19 @@
+
+    long countByArticleId(long articleId);
+
+    Page<Comment> findByArticle_Id(Long articleId, Pageable pageable);
```
countByArticleId는 ArticleId이고 이 메소드는 Article_Id인 이유는?

soojinroh replied:
Repository에서 method를 만들 때 자동완성 기능에 의존해서 네이밍을 하다보니 Article_Id와 같이 네이밍을 했던 것 같습니다. 수정했습니다!

pobiconan commented on src/main/java/com/woowacourse/dsgram/domain/repository/query/HashTagQuery.java:
```diff
@@ -0,0 +1,10 @@
+
+public class HashTagQuery {
+    public static final String FIND_ALL_BY_QUERY =
+            "SELECT new com.woowacourse.dsgram.domain.HashTagSearchResult(h.keyword, count(h)) " +
```
이와 같이 하드코딩할 경우 패키지 rename 리팩토링을 하면 문제가 되지 않을까?
이 문제를 해결할 수 있는 다른 방법은 없나?

ep1stas1s replied:
패키지 name에 의존하지 않도록 수정하였습니다

pobiconan commented on src/main/java/com/woowacourse/dsgram/service/ArticleService.java:
```diff
@@ -75,4 +91,53 @@ public void delete(long articleId, LoggedInUser loggedInUser) {
+                .collect(toList());
+    }
+
+    public FeedInfo getFeedInfo(String fromNickName, String toNickName) {
```
이 메소드 하나 때문에 많은 Service를 DI해야 한다.
이 메소드를 별도의 Service로 분리하는 것과 지금과 같이 구현하는 것 중 어느 것이 더 나을까?

imkimheejoo replied:
많은 서비스 DI를 피하기 위해 FeedService 라는 새로운 서비스를 만들었습니다. 
변경된 코드에서는 feedInfo에 article을 사용하지 않고 새 요청을 보내 Article 정보를 가져와 현재 FeedInfo에는 팔로우정보들만 사용되는 것을 확인되어 article 정보를  가져오는 로직을 제거했습니다!

pobiconan commented on src/main/java/com/woowacourse/dsgram/service/CommentService.java:
```diff
@@ -0,0 +1,75 @@
+    private static final int BLANK_CONTENTS = 0;
+    private CommentRepository commentRepository;
+    private ArticleRepository articleRepository;
+    private UserRepository userRepository;
```
ArticleService에서는 UserService와 의존관계를 맺고 있다.
이곳에서는 UserService가 아닌 UserRepository와 의존관계를 맺고 있는 이유는?

soojinroh replied:
UserRepository를 이용해서 findById등의 간단한 작업만을 처리하다보니 UserService를 의존하고 메서드를 만들어 사용하는 게 비용이라고 느껴졌는데

1. useRrepository.findById와 같은 간단한 로직이라도 여러 service에서 중복 사용되다 보니 UserService의 findByUserId 메서드를 사용해서 중복을 제거할 수 있고
2. UserService에서 userRepository가 아닌 다른 Repository를 직접 의존해서 사용하는 것이 좋지 않다고 느껴져서 UserRepository가 아닌 UserService를 의존하는 것으로 의존관계를 수정하였습니다. 하지만 정확히 어떤 이유에서 다른 Repository와 의존관계를 맺는 게 좋지 않다고 설명할 수 있을 정도로 파악하고 있지는 못해서 이부분에 대해서는 좀 더 고민해보겠습니다.

pobiconan commented on src/main/java/com/woowacourse/dsgram/service/dto/follow/FollowRelation.java:
```diff
@@ -0,0 +1,18 @@
+    FOLLOW, UNFOLLOW, MINE;
+
+    public static FollowRelation getRelation(Follow follow, User guest, User feedOwner) {
+        if (guest.equals(feedOwner)) {
```
`guest.isSameUser(feedOwner)`는 어떨까?

os94 replied:
수정했습니다! 가독성 👍

pobiconan commented on src/main/java/com/woowacourse/dsgram/web/controller/ArticleApiController.java:
```diff
@@ -2,44 +2,64 @@
+    @GetMapping
+    public ResponseEntity showArticles(@UserSession LoggedInUser loggedInUser) {
+        List<Article> articles = articleService.getArticlesByFollowings(loggedInUser.getNickName());
+        List<ArticleInfo> articleInfos = articles.stream().map(article -> ArticleAssembler.toArticleInfo(article)).collect(Collectors.toList());
```
Domain -> Dto 변환 작업을 이곳에서 하고 있다.
Service에서 하는 것과 Controller 중 어느 곳에서 구현하는 것이 나을까?

jinwookh replied:
현재로서는 로직이 간단하기에 controller에 넣어도 가독성은 큰 문제가 없다고 생각합니다. 
하지만 컨트롤러를 최대한 간단하게 만들자는 컨벤션을 지키기 위해, 그리고 다른 컨트롤러들과 통일성을 유지하기 위해 dto 변환작업을 서비스로 옮겼습니다!

pobiconan commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,133 @@
+insert into hash_tag(keyword, article_id, created_date, modified_date)
+values ('#hello', 23, CURRENT_TIME(), CURRENT_TIME());
+insert into hash_tag(keyword, article_id, created_date, modified_date)
+values ('#hello', 24, CURRENT_TIME(), CURRENT_TIME());
```
테스트 데이터 추가 👍 
하지만 이런 반복적인 데이터라면 sql 쿼리보다 프로그램을 통해 추가하는 것이 더 낫지 않을까?

https://spring.io/guides/gs/accessing-data-jpa/ 문서를 보면 서버 시작할 때 테스트 데이터 추가하는 부분과 비슷하게..

jinwookh replied:
프로그램을 통해 추가하는 것으로 바꿨습니다.
수정함으로써, 중복된 쿼리문을 for문법을 통해 간단하게 바꿀 수 있었습니다.

하지만 다음과 같은 문제가 있습니다.
저희는 이미지 파일을 로컬에 저장하기 때문에, 이미지가 잘 나오는지 확인하려면 이미지 경로를 file_info 테이블에 저장해 두어야 합니다.
문제는 static 파일 경로가 운영체제에 따라 달라진다는 점입니다.
운영체제에 따라 달라지는 파일 경로를 처리하는 방법이 필요한 상황입니다.

하지만 해법을 찾지 못했습니다.
결국 이미지 파일을 웹에 저장하여 파일 경로를 웹 url로 만드는 것이 유일한 해법일 것 같습니다.
이  문제는 data.sql로 적용했을 때도 마찬가지로 발생하므로, 피드백을 적용하여 sql 중복을 줄였습니다!

jinwookh replied:
프로그램으로 수정하는 방법으로 바꿨습니다!
data.sql로 바꾸고 나니, 용량이 많은 데이터도 일일히 sql문을 쓰지 않고 for문으로 처리할 수 있었습니다.

배포 때는 테스트 데이터가 불필요하므로 위 메서드를 삭제한 채로 배포해야 하는데, 이를 자동화 할 수 있는지에 대해 고민 중입니다!

pobiconan commented on src/main/resources/static/js/article.js:
```diff
@@ -78,10 +77,27 @@ const ARTICLE_APP = (() => {
         };
 
+        // TODO search-result.js와 중복!!
+        const loadArticles = () => {
```
callback chain이 발생해 코드가 복잡해 지고 있는데
좀 더 깔끔하게 구현할 방법이 없을까?

jinwookh replied:
``` javascript
const handleArticleInfo = articleInfo => {
    cards.insertAdjacentHTML('beforeend', template.card(articleInfo));
    fileLoader.loadMediaFile(fileLoader, articleInfo.articleFileName, articleInfo.articleId);
    fileLoader.loadProfileImageFile(fileLoader, articleInfo.userId, "thumb-img-user-");
};

const handleResponse = data => {
    if (articleCount != null) {
        articleCount.innerText = data.totalElements;
    }
    data.content.forEach(handleArticleInfo);
    headerService.applyHashTag();
};

// TODO search-result.js와 중복!!
const loadArticlesFrom = (url) => {
    return (page) => {
        const addArticle = response => {
            response.json()
                .then(handleResponse);
        }
        connector.fetchTemplateWithoutBody(`${url}?page=${page}`, connector.GET, addArticle);
    };
};
```
메서드를 나누어 가독성을 높였습니다!

pobiconan commented on src/main/java/com/woowacourse/dsgram/service/dto/follow/FollowRelation.java:
```diff
@@ -0,0 +1,18 @@
+public enum FollowRelation {
+    FOLLOW, UNFOLLOW, MINE;
+
+    public static FollowRelation getRelation(Follow follow, User guest, User feedOwner) {
```
이 메소드와 같은 부분이 단위 테스트하기 쉬운 부분인데 테스트가 없는 점 아쉽다.

pobiconan reviewed (changes requested):
3주차 들어서면 상당히 많은 기능을 추가했네요. 💯 
websocket 적용해 채팅 기능구현까지..
코드도 전체적으로 컨벤션을 잘 지키고 깔끔하게 구현 잘 했네요.
단, 아쉬운 점이라면 많은 기능 추가해 비해 이를 테스트하는 테스트 구현이 적은 점이 아쉽네요.
우리가 3주 동안 많은 기능을 추가하는 것이 목표가 아니라 하나를 구현하더라도 절차를 지키면서 구현하는 경험을 하는 것이 목표인데요.

soojinroh committed:
fix: comments (#49)

jinwookh committed:
refactor: data.sql을 application 속 함수로 대체 (#53)

soojinroh committed:
Fix : comment feedback (#50)

* fix: comments

* fix : 피드백 반영

BaseEntity 접근 제어자 수정
CommentService 의존 관계 수정
Comment에서 BaseEntity 상속

imkimheejoo committed:
feat: like (-test..) (#51)

* feat: like 기능 구현

* chore: followInfo 삭제

soojinroh committed:
Fix : apply comment to other page (#56)

* fix: comments

* fix : 다른 페이지에 comment 기능 적용

ep1stas1s committed:
test: hash Tag test 추가 (#57)

수고하셨습니다 :)

wmakerjun commented on src/main/resources/static/js/card.js:
```diff
@@ -76,9 +81,38 @@ const CARD_APP = (() => {
+                'Content-Type': 'application/json; charset=UTF-8',
+                'Accept': 'application/json'
+            };
+            let target = event.target;
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

imkimheejoo replied:
수정했습니다~!

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+
+    const CommentService = function() {
+        const save = event => {
+            let target = event.target;
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+                return;
+            }
+
+            let targetArticleId = target.getAttribute("data-article-id");
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+            }
+
+            let targetArticleId = target.getAttribute("data-article-id");
+            let commentContents = target.parentNode.parentNode.querySelector("textarea").value;
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+
+            const updateDomAfterSaveComment = function(target, res) {
+                target.parentNode.parentNode.querySelector('textarea').value = '';
+                let addedCountOfComment = Number(target.parentElement.parentElement.parentElement.querySelector('span').getAttribute('data-count-comment')) + 1;
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+        };
+
+        const show = event => {
+            let target = event.target;
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+                    ${comment}
+                 {{/comments}}`;
+
+            let commentItemTemplate = Handlebars.compile(commentTemplate);
```
93 lin~97 line, 104 line의 변수들은 
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+        };
+
+        const remove = event => {
+            let target = event.target;
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/room.js:
```diff
@@ -0,0 +1,104 @@
+CHAT_APP = (() => {
```
변수 선언을 위해서 const를 써줘야 하지 않을까요~? :)

wmakerjun commented on src/main/resources/static/js/card.js:
```diff
@@ -76,9 +81,38 @@ const CARD_APP = (() => {
+            };
+            let target = event.target;
+            if (target.tagName === 'I' && target.classList.contains('fa') ) {
+                const articleId = target.closest('.article-card').getAttribute('id');
```
[dataset](https://developer.mozilla.org/ko/docs/Learn/HTML/Howto/%EB%8D%B0%EC%9D%B4%ED%84%B0_%EC%86%8D%EC%84%B1_%EC%82%AC%EC%9A%A9%ED%95%98%EA%B8%B0)이라는 data 속성을 가져오는 메서드가 있기 때문에 그걸 활용하는게 더 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+            };
+
+            const drawComment = function (comment) {
+                let commentTemplate = template.comment(comment.contents, comment.user.nickName, comment.id);
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+            const drawComment = function (comment) {
+                let commentTemplate = template.comment(comment.contents, comment.user.nickName, comment.id);
+
+                let commentList = target.parentNode.parentNode.parentNode.querySelector('.list-info');
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+            const updateDomAfterSaveComment = function(target, res) {
+                target.parentNode.parentNode.querySelector('textarea').value = '';
+                let addedCountOfComment = Number(target.parentElement.parentElement.parentElement.querySelector('span').getAttribute('data-count-comment')) + 1;
+                target.parentElement.parentElement.parentElement.querySelector('span').setAttribute('data-count-comment', addedCountOfComment);
```
parentElement를 chaining해서 element를 찾는 방식은 나중에 저 depth에 다른 ui 엘리먼트 같은게 추가되면서 에러를 발생시킬 확률을 더 높여줄 수 있어요~ 그래서 parentElement 대신에 [closest](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest)를 활용하는걸 추천합니다 :)

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+                return;
+            }
+
+            let commentId = target.getAttribute('data-comment-id');
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+
+            const ifSucceed = (target) => {
+                alert('댓글이 삭제 되었습니다.');
+                let spanTag = target.parentNode.parentNode.parentNode.parentNode.querySelector('span');
```
129~30 line 
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+                target.parentElement.parentElement.remove();
+            };
+
+            connector.fetchTemplateWithoutBody('/api/comments/'+commentId,connector.DELETE,ifSucceed(target));
```
[문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/follow.js:
```diff
@@ -0,0 +1,119 @@
+const FOLLOW_APP = (() => {
+
+    const FollowController = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

imkimheejoo replied:
수정했습니다!

wmakerjun commented on src/main/resources/static/js/follow.js:
```diff
@@ -0,0 +1,119 @@
+        }
+    };
+
+    const FollowService = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

imkimheejoo replied:
수정했습니다. ㅎㅎ

wmakerjun commented on src/main/resources/static/js/header.js:
```diff
@@ -28,63 +28,70 @@ HEADER_APP = (() => {
 
     const HeaderService = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/myFeed.js:
```diff
@@ -0,0 +1,70 @@
+const MY_FEED_APP = (() => {
+    'use strict';
+
+    const MyFeedController = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/myFeed.js:
```diff
@@ -0,0 +1,70 @@
+
+    };
+
+    const MyFeedService = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/room.js:
```diff
@@ -0,0 +1,104 @@
+
+    let stompClient = null;
+
+    const ChatController = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/room.js:
```diff
@@ -0,0 +1,104 @@
+        }
+    };
+
+    const ChatService = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/search-result.js:
```diff
@@ -0,0 +1,59 @@
+const SEARCH_APP = (() => {
+
+    const SearchController = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/search-result.js:
```diff
@@ -0,0 +1,59 @@
+        }
+    };
+
+    const SearchService = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/template.js:
```diff
@@ -0,0 +1,174 @@
+TEMPLATE_APP = (() => {
+    const TemplateService = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

imkimheejoo replied:
수정했습니다~

wmakerjun commented on src/main/resources/static/js/user.js:
```diff
@@ -150,7 +150,15 @@ const USER_APP = (() => {
             const connector = FETCH_APP.FetchApi();
 
             const getImage = response => {
+                function checkEmptyImage(buffer) {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/templates/article.html:
```diff
@@ -15,28 +16,50 @@
-            fileLoader.loadMediaFile(fileLoader, [[${article.fileInfo.fileName}]], [[${article.id}]]);
-        };
-        window.addEventListener('DOMContentLoaded', fileLoad);
+        (function() {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/templates/article.html:
```diff
@@ -15,28 +16,50 @@
+                    });
+                });
+
+            const articleFileLoad = function (articleFileName, articleId) {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/templates/article.html:
```diff
@@ -15,28 +16,50 @@
+                fileLoader.loadMediaFile(fileLoader, articleFileName, articleId);
+            };
+
+            const userFileLoad = function (userFileName, userId) {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/article.js:
```diff
@@ -3,26 +3,24 @@ const ARTICLE_APP = (() => {
 
     const ArticleController = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/article.js:
```diff
@@ -31,6 +29,12 @@ const ARTICLE_APP = (() => {
     };
 
     const ArticleService = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/card.js:
```diff
@@ -3,19 +3,24 @@ const CARD_APP = (() => {
     const CardController = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+                contents: commentContents,
+            };
+
+            const drawComment = function (comment) {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/observer.js:
```diff
@@ -0,0 +1,34 @@
+const OBSERVER_APP = (() => {
+    'use strict';
+
+    const observeService = function () {
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 여기서도 다른 부분과 동일하게 arrow function으로 맞춰주면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/follow.js:
```diff
@@ -0,0 +1,119 @@
+
+        const follow = event => {
+            event.preventDefault();
+            const ifSucceed = () => window.location.href = '/users/' + document.getElementById('feedOwner').innerText;
```
[문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

imkimheejoo replied:
수정했습니다~

wmakerjun commented on src/main/resources/static/js/follow.js:
```diff
@@ -0,0 +1,119 @@
+            };
+
+            const nickName = document.querySelector('#feedOwner').innerHTML;
+            connector.fetchTemplateWithoutBody('/followers/' + nickName, connector.GET, ifSucceed);
```
[문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

imkimheejoo replied:
수정했습니다!

wmakerjun commented on src/main/resources/static/js/follow.js:
```diff
@@ -0,0 +1,119 @@
+            };
+
+            const nickName = document.querySelector('#feedOwner').innerHTML;
+            connector.fetchTemplateWithoutBody('/followings/' + nickName, connector.GET, ifSucceed);
```
[문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

imkimheejoo replied:
수정했습니다!!

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+                connector.GET,
+                (response) => {
+                    response.json().then(res => {
+                        let data = {comments: res.content};
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,157 @@
+                return;
+            }
+
+            let page = parseInt((countOfBrowserComments / 5).toString());
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun commented on src/main/resources/static/js/template.js:
```diff
@@ -0,0 +1,174 @@
+        const searchResult = hashTag => {
+            const template =
+                `<li class="search-result-item">
+                                    <div class="row align-items-center margin-10">
```
js파일에서도 template 부분 indent 컨벤션을 같이 맞춰주면 더 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/template.js:
```diff
@@ -0,0 +1,174 @@
+        const card = articleInfo => {
+            const template =
+                `<div id="${articleInfo.articleId}" class="article-card card widget-feed no-pdd mrg-btm-70 shadow-sm">
+                                    <div class="feed-header padding-15">
```
js파일에서도 template 부분 indent 컨벤션을 같이 맞춰주면 더 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/template.js:
```diff
@@ -0,0 +1,174 @@
+        };
+
+        const heartState = (toggle) => {
+            if(toggle) {
```
if 다음에 괄호 사이에 [공백컨벤션](https://ui.toast.com/fe-guide/ko_CODING-CONVENSION/#%EA%B3%B5%EB%B0%B1)에 맞게 공백을 넣어주세요 :)

imkimheejoo replied:
안쓰는 로직이여서 삭제했습니다~

wmakerjun commented on src/main/resources/static/js/template.js:
```diff
@@ -0,0 +1,174 @@
+            if (message.from.id !== sessionUserId) {
+                template =
+                    `<div class="incoming_msg">
+                <div class="incoming_msg_img">
```
js파일에서도 template 부분 indent 컨벤션을 같이 맞춰주면 더 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/template.js:
```diff
@@ -0,0 +1,174 @@
+        }
+
+        const chatMessage = (message, sessionUserId) => {
+            let template;
```
변수를 재사용할 일이 없다면 변수에 담지 말고 그냥 바로 html 부분을 바로 return 하면 어떨까요~?
```
return `<div></div>`
```

wmakerjun reviewed (changes requested):
구조화 과정이 돋보이는 프론트 코드였습니다 :) 여기까지 만드느라 너무너무 고생많았습니다 ㅎㅎ 💯💯 
최종 마무리하기전에 피드백은 코드가 정돈될 수 있게 컨벤션 위주로만 달았으니, 해당 부분 리팩토링 할 때 참고해주세요  :smile
남은 시간도 화이팅입니닷~!

jinwookh committed:
feat: infinite scroll in index (#58)

* refactor: article.js

메서드 분리

* refactor: index.js

index페이지에서 article들을 가져오는 로직을 article.js-> index.js로 분리

* refactor: article.html

fileLoader 메서드를 사용하도록 리팩토링

* feat: index 페이지에서 무한 스크롤링 구현

* feat: test data에서 hashtag contents 수정

* refactor: articles 가져오는 함수 중복 제거

articles 가져오는 함수는 이제 articleService.loadArticles(url)을 이용하면 됩니다.

* refactor: default-without-script layout 삭제

default layout 속 addScript fragment로 개별 script문 추가

* refactor: showArticles로직을 컨트롤러에서 서비스로 옮김

포비 피드백 반영

* refactor: test data 중 file 경로부분을 상대 경로로 구현

* refactor: 공통 script인 script.html에 handlebar.js  추가

ep1stas1s committed:
fix: CSS & XSS (#59)

* fix: dm css

* fix: dm list css 수정

* fix: random nickname

* fix: dm list css

* feat: xss 차단

* refactor: data.sql 제거

* refactor: query 수정

* refactor: 상수

jinwookh committed:
add : Jenkinsfile (#61)

os94 committed:
feat : 파일 확장자 검증로직 추가 (#63)

* refactor : build.gradle의 h2database를 testRuntimeOnly 속성으로 변경

* refactor : User.equals() -> User.isSameUser()

* test : FollowRelationTest 추가

* feat : 파일 확장자 검증로직 추가

ep1stas1s committed:
fix: 오기 수정 (#64)

imkimheejoo committed:
feat: 좋아요 한 사람들 리스트 모달 (#60)

* feat: like 기능 구현

* chore: followInfo 삭제

* fix: 이미 좋아요 상태일 때 하트상태 불일치 수정

* feat: 좋아요 리스트 Modal 구현

* fix: like uri 수정 및 js 컨벤션 통일

* fix: jun feedback 반영

* fix: jun feedback arrow 컨벤션 반영

* refactor : Rename FollowInfo -> UserInfo

soojinroh committed:
Fix : bug fix 및 feedback 적용 (#65)

* fix: comments

* fix : CommentControllerTest create테스트 수정

* fix : 버그리포팅 댓글 4번 디버그

* fix : 피드백 반영 및 버그 픽스

* merge : resolve conflict

os94 committed:
fix : Controller, Service의 arrow fn -> function()방식으로 통일

os94 committed:
fix : 게시글 수정 버그 및 Bean 순환참조 이슈 수정 (#66)

* fix : merge conflict 해결 및 ArticleService에서 CommentRepository대신 CommentService를 주입받도록 수정

* fix : Bean 순환참조 이슈로 ArticleService, CommentService를 멤버변수 주입으로 수정

* fix : 게시글 수정이 되지않는 버그 수정

imkimheejoo committed:
refactor: feedInfo service 분리 (#67)

* refactor: getFeedInfo 메서드 FeedService 로 분리

* refactor: 안쓰는 로직 제거 및 접근제어자 수정

imkimheejoo committed:
test: like Test코드 추가 (#68)

* refactor: getFeedInfo 메서드 FeedService 로 분리

* refactor: 안쓰는 로직 제거 및 접근제어자 수정

* test: like test 구현

soojinroh committed:
fix : copyUrl 버튼 적용 및 ControllerServiceTest 코드 추가 (#69)

* fix : copyUrl 원상복구

* test : CommentServiceTest 테스트 코드 추가

os94 committed:
refactor : Service의 @Transactional 사용방식 통일 (#70)

* refactor : Service의 @Transactional 사용방식 통일

jinwookh committed:
refactor: apache-commons-text library를 compileOnly -> implementation으로 변경

초기 더미 데이터들 넣는 메서드도 삭제

imkimheejoo committed:
test: like 목록 조회 (#71)

* refactor: getFeedInfo 메서드 FeedService 로 분리

* refactor: 안쓰는 로직 제거 및 접근제어자 수정

* test: like test 구현

* test: 좋아요 리스트 조회 Test 추가

imkimheejoo committed:
refactor: like 기능 리팩토링 (#72)

* refactor: getFeedInfo 메서드 FeedService 로 분리

* refactor: 안쓰는 로직 제거 및 접근제어자 수정

* test: like test 구현

* test: 좋아요 리스트 조회 Test 추가

* refactor: follow 기능 url 이름 변경

* refactor: like 기능 리팩토링

* test: 좋아요 테스트 수정

* chore: 주석제거

ep1stas1s committed:
feat: rest docs (#73)

* refactor: getFeedInfo 메서드 FeedService 로 분리

* refactor: 안쓰는 로직 제거 및 접근제어자 수정

* test: like test 구현

* refactor: hash tag path 변경

* feat: hash tag rest docs

* feat: user rest docs

* feat: article rest docs

* refactor: docs 수정

* test: 좋아요 리스트 조회 Test 추가

* refactor: follow 기능 url 이름 변경

* test: test 수정

* feat: like rest docs

* feat: comment rest docs

* refactor: like rest docs 수정

* feat: follow rest docs

* refactor: rest docs 양식 수정

* refactor: path 변경

soojinroh committed:
test : CommentServiceTest 테스트 코드 추가 (#74)

수고하셨습니다 :)

imkimheejoo committed:
fix: getCountOfComment 메서드 이름 변경

imkimheejoo committed:
chore: todo 제거

os94 committed:
refactor : 프론트 수정 (#75)

* fix : 게시글 내용에 대한 css 수정

* refactor : 프론트의 DM목록 띄우는 방식을 변경

os94 committed:
feat : 일부 url 매핑 연결

os94 committed:
fix : spring.jpa.hibernate.ddl-auto 설정을 none -> create로 변경

woowahan-pjs reviewed (approved):
축하합니다! 백엔드 Java 레벨 2의 모든 미션 완료하였어요. 🥇 
나만의 블로그부터 미니 프로젝트까지
수많은 미션을 수행하느라 대단히 고생 많으셨습니다. 💯 
좋은 개발자가 되기 위한 소중한 밑거름이 되었길 바랍니다.
이만 Merge 하겠습니다.
